docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+  
+**Note:** Flows related to CET support inside enclave will be considered as a future enhancement.
+
+Future kernel may extend mmap and mprotect syscalls to support SGX EDMM usages. But we can't add/change syscall interfaces from an out-of-tree driver.  So, in this proposal for possible driver implementation, we reuse mmap for dynamic enclave memory mapping and expose a new IOCTL, sgx_enclave_mprotect, for enclave page modification. 
+ 
+## mmap
+
+After enclave is initialized (EINIT IOCTL done), the standard Linux mmap syscall can be used to create a new mapping configured for dynamically allocating enclave memory using EAUG. Following comments are specific to SGX EDMM usages, please refer to [mmap man page](https://man7.org/linux/man-pages/man2/mmap.2.html) for generic definitions.  
+
+### Remarks
+
+- To create a mapping for dynamic enclave memory allocation, mmap must be called with an open enclave file descriptor and with PROT_READ | PROT_WRITE for protection flags. 
+  - Enclave must issue EACCEPT for the pages after mmap, then enclave may modify the content of the pages and extend/reduce permissions in secure way.
+- The offset in mmap parameter must be zero for enclaves.
+- MAP_* flags must be MAP_SHARED  | MAP_FIXED masked with optional flags:
+    - MAP_POPULATE: hint for kernel to EAUG pages as soon as possible.  
+    - MAP_GROWSDOWN: used for stacks.  The mapping will grow down to the next mapping.
+- If and only if the address range are within the ELRANGE of the enclave associated with the file descriptor, the mapping will be created. However, user space should not expect EAUG be done by the mmap call.
+  - The kernel can choose EAUG pages immediately (likely for MAP_POPULATE), or EAUG pages upon page faults within the VMA, similar to how kernel would allocate regular memory.
+- The kernel will assume the newly requested mapping is for dynamic allocation and initial permissions must be RW until user space request changes later. 
+
+**Implementation Notes:** Current [SGX kernel patches](https://patchwork.kernel.org/project/intel-sgx/patch/20201112220135.165028-11-jarkko@kernel.org/) limit PTE permissions to the EPCM permissions given in SEC_INFO during EADD IOCTL calls. The dynamic allocation mappings should not be subject to those limits. A possible implementation may have these changes:
+  - sgx_encl_may_map 
+    - enforces RW permissions for pages other than those loaded due to EADD or ECREAT.
+    - set up flags to track dynamic pages: type, permissions flag
+  - sgx_vma_mprotect
+    - Allow permissions changes to dynamic pages within limitations of OS policies, e.g.,
+	    - never allow WX
+	    - SELinux policy specific to SGX enclaves
+    - update flags for the dynamic pages
+
+## munmap 
+Calling munmap on an enclave page (dynamic allocated or not) has exactly the same effect of calling munmap on a regular RAM page. No sgx specific interface is needed. No behavior changes to current kernel space implementation.
```

Comment:
I don't understand this part. The `munmap()` logic must be changed in two components:
1. The SGX driver must perform `PT_TRIM`.
2. The trusted SGX runtime (e.g. Graphene) must perform `EACCEPT` after the driver's `PT_TRIM`.

So what does it mean "no sgx specific interface is needed"? Not needed in which components?